File: python/nameq.py
"""Application support library for using information exported by nameq."""

from __future__ import absolute_import

__all__ = [
	"DEFAULT_STATEDIR",
	"Feature",
	"FeatureMonitor",
	"log",
]

import errno
import fcntl
import json
import logging
import os
import select
import threading

import inotify

DEFAULT_STATEDIR = "/run/nameq/state"

log = logging.getLogger("nameq")

class Feature(object):
	"""Represents a momentary state of a feature on a host.

	   Has properties 'name' (of the feature), 'host' (IPv4 or IPv6 address
	   string of the host where the feature exists) and 'value' (None if
	   feature was removed)."""

	def __init__(self, name, host, value=None):
		self.name = name
		self.host = host
		self.value = value

	def __repr__(self):
		return "Feature(name=%r, host=%r, value=%r)" % (self.name, self.host, self.value)

class _FeatureMonitor(object):

	def __init__(self, statedir=DEFAULT_STATEDIR):
		featuredir = os.path.join(statedir, "features")

		try:
			os.mkdir(featuredir, 0o755)
		except OSError as e:
			if e.errno != errno.EEXIST:
				raise

		self._featuredir = os.path.realpath(featuredir)

		self._queued_features = []

		self._fd = inotify.init(inotify.CLOEXEC|inotify.NONBLOCK)
		ok = False
		try:
			inotify.add_watch(self._fd, self._featuredir, inotify.ONLYDIR|inotify.CREATE|inotify.DELETE|inotify.DELETE_SELF)
			ok = True
		finally:
			if not ok:
				os.close(self._fd)

		try:
			featurenames = os.listdir(self._featuredir)
		except Exception as e:
			log.exception("listing %s", self._featuredir)
		else:
			for featurename in featurenames:
				featurepath = os.path.join(self._featuredir, featurename)
				if self._add_feature(featurepath):
					try:
						hostnames = os.listdir(featurepath)
					except Exception as e:
						log.exception("listing %s", featurepath)
					else:
						for hostname in hostnames:
							self._add_host(os.path.join(featurepath, hostname))

	def __enter__(self):
		return self

	def __exit__(self, *exc):
		self.close()

	def close(self):
		log.debug("_FeatureMonitor.close method not implemented")

	def _handle(self, event):
		if event.mask & inotify.CREATE:
			self._add_feature(event.name)

		if event.mask & inotify.DELETE:
			if os.path.dirname(event.name) == self._featuredir:
				self._remove_feature(event.name)
			else:
				self._remove_host(event.name)

		if event.mask & inotify.DELETE_SELF:
			self.close()

		if event.mask & inotify.MOVED_TO:
			self._add_host(event.name)

	def _add_feature(self, path):
		try:
			inotify.add_watch(self._fd, path, inotify.ONLYDIR|inotify.DELETE|inotify.MOVED_TO)
			return True
		except Exception:
			log.exception("adding watch for %s", path)
			return False

	def _remove_feature(self, path):
		try:
			inotify.rm_watch(self._fd, path)
		except Exception:
			log.exception("removing watch for %s", path)

	def _add_host(self, path):
		try:
			f = open(path)
		except Exception:
			return

		try:
			with f:
				data = f.read()
		except Exception:
			log.exception("reading %s", path)
			return

		self._enqueue_feature(path, data)

	def _remove_host(self, path):
		if os.access(path, os.F_OK):
			return

		self._enqueue_feature(path)

	def _enqueue_feature(self, path, data=None):
		value = None

		if data is not None:
			try:
				value = json.loads(data)
			except Exception:
				log.exception("decoding JSON data %r", data)
				return

		self._queued_features.append(Feature(
			name = os.path.basename(os.path.dirname(path)),
			host = os.path.basename(path),
			value = value,
		))

class FeatureMonitor(_FeatureMonitor):
	"""Watches the nameq runtime state for changes.  Either the 'changed' method
	   must be implemented in a subclass, or a callable must be provided as the
	   'changed' parameter.  It will be invoked with a Feature instance, or the
	   terminator when the monitor is closed.  The state directory must
	   exist."""

	_bufsize = 65536

	def __init__(self, changed=None, terminator=None, statedir=DEFAULT_STATEDIR):
		super(FeatureMonitor, self).__init__(statedir)

		if changed is not None:
			self.changed = changed

		self.terminator = terminator

		self._pipe = os.pipe()

		try:
			os.O_CLOEXEC
		except AttributeError:
			pass
		else:
			for fd in self._pipe:
				flags = fcntl.fcntl(fd, fcntl.F_GETFD)
				fcntl.fcntl(fd, fcntl.F_SETFD, flags|os.O_CLOEXEC)

		flags = fcntl.fcntl(self._pipe[0], fcntl.F_GETFL)
		fcntl.fcntl(self._pipe[0], fcntl.F_SETFL, flags|os.O_NONBLOCK)

		self._thread = threading.Thread(target=self._loop)
		self._thread.start()

	def close(self):
		"""Stop watching and invoke the callback with the terminator."""

		os.close(self._pipe[1])
		self._thread.join()

	def changed(self, feature):
		log.debug("FeatureMonitor.changed method not implemented")

	def _iter(self):
		while True:
			try:
				readable, _, _ = select.select([self._fd, self._pipe[0]], [], [])
			except select.error as e:
				if e.args[0] == errno.EINTR:
					continue
				raise

			if self._fd in readable:
				try:
					buf = os.read(self._fd, self._bufsize)
				except OSError as e:
					if e.errno != errno.EAGAIN:
						raise
				else:
					assert buf
					for event in inotify.unpack_events(buf):
						yield event

			if self._pipe[0] in readable:
				try:
					os.read(self._pipe[0], 1)
				except OSError as e:
					if e.errno != errno.EAGAIN:
						raise
				else:
					break

	def _loop(self):
		try:
			self._deliver()
			for event in self._iter():
				self._handle(event)
				self._deliver()
		finally:
			self.changed(self.terminator)

	def _deliver(self):
		for feature in self._queued_features:
			try:
				self.changed(feature)
			except Exception:
				log.exception("uncaught exception in FeatureMonitor.changed callback")

		del self._queued_features[:]

File: go/nameq.go
// Package nameq is an application support library for using the information
// exported by the nameq service.
package nameq

import (
	"io/ioutil"
	"net"
	"os"
	"path/filepath"
	"strings"
	"syscall"

	"github.com/fsnotify/fsnotify"
)

const (
	// The directory used when an empty string is given to SetFeature or RemoveFeature.
	DefaultFeatureDir = "/etc/nameq/features"

	// The directory used when an empty string is given to NewFeatureMonitor.
	DefaultStateDir = "/run/nameq/state"
)

// SetFeature adds or updates a local feature.  data must be a valid JSON
// document.  DefaultFeatureDir is used if an empty string is given.  Redundant
// calls are ok.
func SetFeature(featureDir, name string, data []byte) error {
	if featureDir == "" {
		featureDir = DefaultFeatureDir
	}

	if len(data) == 0 {
		panic("no data for feature")
	}

	return createConfigFile(featureDir, name, data)
}

// RemoveFeature removes a local feature.  DefaultFeatureDir is used if an
// empty string is given.  Redundant calls are ok.
func RemoveFeature(featureDir, name string) error {
	if featureDir == "" {
		featureDir = DefaultFeatureDir
	}

	return removeConfigFile(featureDir, name)
}

func createConfigFile(dir, name string, data []byte) (err error) {
	if err = os.MkdirAll(dir, 0755); err != nil {
		return
	}

	tmpDir := filepath.Join(dir, ".tmp")
	os.Mkdir(tmpDir, 0700)

	tmpPath := filepath.Join(tmpDir, name)

	if err = ioutil.WriteFile(tmpPath, data, 0644); err != nil {
		return
	}

	path := filepath.Join(dir, name)

	if err = os.Rename(tmpPath, path); err != nil {
		os.Remove(tmpPath)
	}
	return
}

func removeConfigFile(dir, name string) (err error) {
	if err = os.Remove(filepath.Join(dir, name)); err != nil {
		if err.(*os.PathError).Err == syscall.ENOENT {
			err = nil
		}
	}
	return
}

// Feature represents a momentary state of a feature on a host.
type Feature struct {
	Name string // Name of the feature.
	Host net.IP // IPv4 or IPv6 address of the host where the feature exists.
	Data []byte // JSON value if feature added or updated, or nil if removed.
}

// String returns the feature name.
func (f *Feature) String() string {
	return f.Name
}

// Logger is a subset of the standard log.Logger.
type Logger interface {
	Print(v ...interface{})
}

// FeatureMonitor watches the nameq runtime state for changes.
type FeatureMonitor struct {
	// Produces current information, followed by updates in real time.  There
	// may be seemingly redundant entries when a feature is updated rapidly.
	C <-chan *Feature

	// Boot is closed after all existing features have been delivered via C.
	// Client code is free to set this member to nil after it has been closed.
	Boot <-chan struct{}

	logger  Logger
	closed  chan struct{}
	watcher *fsnotify.Watcher
	queued  []*Feature
}

// NewFeatureMonitor watches the specified state directory, or the default
// state directory if an empty string is given.  The directory must exist.  The
// logger is used for I/O errors, unless nil.
func NewFeatureMonitor(stateDir string, logger Logger) (m *FeatureMonitor, err error) {
	if stateDir == "" {
		stateDir = DefaultStateDir
	}

	featureDir := filepath.Join(stateDir, "features")

	if err = os.Mkdir(featureDir, 0755); err != nil {
		if info, statErr := os.Stat(featureDir); statErr != nil || !info.IsDir() {
			return
		}
		err = nil
	}

	if featureDir, err = filepath.Abs(featureDir); err != nil {
		return
	}

	if featureDir, err = filepath.EvalSymlinks(featureDir); err != nil {
		return
	}

	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		return
	}

	if err = watcher.Add(featureDir); err != nil {
		watcher.Close()
		return
	}

	c := make(chan *Feature)
	boot := make(chan struct{})

	m = &FeatureMonitor{
		C:       c,
		Boot:    boot,
		logger:  logger,
		closed:  make(chan struct{}, 1),
		watcher: watcher,
	}

	if infos, err := ioutil.ReadDir(featureDir); err == nil {
		for _, info := range infos {
			m.addFeature(filepath.Join(featureDir, info.Name()))
		}
	} else {
		m.log(err)
	}

	m.queued = append(m.queued, nil) // indicates end of boot

	go m.watchLoop(c, boot, featureDir)

	return
}

// Close stops watching immediately and closes the channel.
func (m *FeatureMonitor) Close() {
	select {
	case m.closed <- struct{}{}:
	default:
	}
}

func (m *FeatureMonitor) watchLoop(c chan<- *Feature, boot chan<- struct{}, featureDir string) {
	defer m.watcher.Close()
	defer close(c)

	// Stripping this prefix from watched names converts them into relative paths
	namePrefixLen := len(featureDir) + 1 // separator

	for {
		input := m.watcher.Events
		output := c
		var outstanding *Feature

		if len(m.queued) > 0 {
			outstanding = m.queued[0]
			input = nil

			if outstanding == nil { // end of boot indicator
				close(boot)
				m.queued = m.queued[1:]
				continue
			}
		} else {
			output = nil
		}

		select {
		case e := <-input:
			if e.Name == featureDir {
				// First level
				if e.Op&fsnotify.Remove != 0 {
					return
				}
			} else {
				relName := e.Name[namePrefixLen:]
				if !strings.ContainsRune(relName, filepath.Separator) {
					// Second level
					m.addFeature(e.Name)
				} else {
					// Deeper level
					hostname := filepath.Base(e.Name)
					if e.Op&fsnotify.Create != 0 {
						m.addHost(hostname, e.Name)
					}
					if e.Op&fsnotify.Remove != 0 {
						m.removeHost(hostname, e.Name)
					}
				}
			}

		case err := <-m.watcher.Errors:
			m.log(err)

		case output <- outstanding:
			m.queued = m.queued[1:]

		case <-m.closed:
			return
		}
	}
}

func (m *FeatureMonitor) addFeature(dir string) {
	if err := m.watcher.Add(dir); err != nil {
		m.log(err)
		return
	}

	infos, err := ioutil.ReadDir(dir)
	if err != nil {
		m.log(err)
		return
	}

	for _, info := range infos {
		m.addHost(info.Name(), filepath.Join(dir, info.Name()))
	}
}

func (m *FeatureMonitor) addHost(hostname, path string) {
	host := m.parseHost(hostname, path)
	if host == nil {
		return
	}

	// File may not exist if we were too slow; just skip this one silently, the
	// delete event will follow.  It will appear spurious for the client code,
	// but that's okay.

	file, err := os.Open(path)
	if err != nil {
		return
	}
	defer file.Close()

	data, err := ioutil.ReadAll(file)
	if err != nil {
		m.log(err)
		return
	}

	m.queued = append(m.queued, &Feature{
		Name: filepath.Base(filepath.Dir(path)),
		Host: host,
		Data: data,
	})
}

func (m *FeatureMonitor) removeHost(hostname, path string) {
	host := m.parseHost(hostname, path)
	if host == nil {
		return
	}

	// File may exist if we were too slow... see the comment in addHost.

	if _, err := os.Stat(path); err == nil {
		return
	}

	m.queued = append(m.queued, &Feature{
		Name: filepath.Base(filepath.Dir(path)),
		Host: host,
	})
}

func (m *FeatureMonitor) parseHost(name string, path string) (host net.IP) {
	if host = net.ParseIP(name); host == nil {
		m.log("unable to parse filename: ", path)
	}

	return
}

func (m *FeatureMonitor) log(args ...interface{}) {
	if m.logger != nil {
		m.logger.Print(args...)
	}
}
